subprojects/build-operations/src/main/java/org/gradle/internal/operations/BuildOperationAncestryTracker.java

```diff
@@ -0,0 +1,29 @@
+/*
+ * Copyright 2020 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.operations;
+
+import javax.annotation.Nullable;
+import java.util.Optional;
+import java.util.function.Function;
+import java.util.function.Predicate;
+
+@SuppressWarnings("Since15")
+public interface BuildOperationAncestryTracker {
```

Comment:
Can you add a `@ServiceScope` annotation here?

Reply:
I can, but I have an issue with how we do `@ServiceScope` now. Not having checks is error-prone, and I don't think the annotation belongs on the implementation class at all.

Reply:
This is the interface, how is it relevant that in your opinion the annotation does not belong on the implementation class?

What is the current approach to `@ServiceScope`? I was under the impression that we should add the annotation to all services. Is this not what we should do?

Reply:
What I meant is that annotating a type is not the right thing to do. (Interface or implementation doesn't matter, but FTR this was the implementation type when I placed the comment. :p)

A service might have multiple instances registered across different scopes for example. Take the fingerprinter services that we have both globally and on a per-project scope. Same for the `WorkExecutor`. You can't capture that with a single annotation, and putting multiple ones on a single type would be confusing.

Then there's the fact that we don't have checks AFAIK; at least forgetting the annotation is not punished in any way. So the coverage is spotty and potentially incorrect/misleading.

Then there's also the question of using these services in a context outside our `ServiceRegistry` DI framework, for example in the Maven cache extension. There `@ServiceScope` means nothing, and could even contradict reality.

We even had to export `@ServiceScope` to non-Gradle consumers, which is a very direct code smell.